src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java

```diff
@@ -1228,17 +1233,31 @@ else if (!referencePath.isDirectory())
                     perModuleArgs.add(sModulePath);
                     StandardCredentials finalCred = cred;
                     URIish finalUrIish = urIish;
-                    executorService.submit(() -> {
+
+                    launchCommands.add(() -> {
                         try {
-                            launchCommandWithCredentials(perModuleArgs, workspace, finalCred, finalUrIish, timeout);
+                            return launchCommandWithCredentials(perModuleArgs, workspace, finalCred, finalUrIish, timeout);
                         } catch (InterruptedException e) {
                             throw new GitException("Interrupted while updating submodule for " + sModuleName);
                         }
                     });
                 }
 
+                List<Future<String>> futures = executorService.invokeAll(launchCommands);
                 executorService.shutdown();
-                executorService.awaitTermination(Long.MAX_VALUE, TimeUnit.NANOSECONDS);
+                try {
+                    executorService.awaitTermination(Long.MAX_VALUE, TimeUnit.NANOSECONDS);
+                } catch (InterruptedException e) {
+                    throw new GitException(e);
+                }
+
+                futures.forEach((future) -> {
+                    try {
+                        future.get();
+                    } catch (Exception e) {
+                        throw new GitException(e);
```

Comment:
Have a look at [ExecutionException](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutionException.html) how to get to original exception in case a thread threw one.

Reply:
Ok, let's collapse this to `catch (ExecutionException | InterruptedException e)`.